### ISSUE TITLE
Disposables should implement finalizer

### ISSUE THREAD

**Issue by Timovzl** [aws/aws-sdk-net#1865]
Many commenters [here](https://github.com/aws/aws-sdk-net/issues/152) report **running out of available connections because they forgot to dispose a resource**. While this could be considered our own fault, it is a common scenario, as evidenced by the comments and by the existence of finalizers.

## Describe the Feature

Any disposable resource that has the potential of keeping a connection open should implement a finalizer.

This way, if disposal is overlooked, resources are merely released _later than necessary_, instead of _never_.

## Proposed Solution

Finalizers are a well-documented feature, used in many places, such as ADO.NET.

### Describe alternatives you've considered

An exception would be _far_ better than the code getting stuck. Yes, it actually gets stuck. Results include jobs hanging indefinitely.

Ideally, both solutions are implemented. That way, if people run out of connections due to other circumstances, they get a clear exception.

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change  

---

This is a :rocket: Feature Request

**Comment by ashishdhingra** [comment]
Refer https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/finalizers#using-finalizers-to-release-resources. Finalizers could also cause performance issues when used incorrectly.

This needs review with the team.

**Comment by Timovzl** [comment]
@ashishdhingra From the doc you shared, I would emphasize the following:

> However, when your application encapsulates unmanaged resources, such as windows, files, and network connections, you should use finalizers to free those resources.

This is currently missing.

> If your application is using an expensive external resource, we also recommend that you provide a way to explicitly release the resource before the garbage collector frees the object. To release the resource, implement a Dispose method from the IDisposable interface that performs the necessary cleanup for the object.

This is already in place.

As for performance impact, it's good to note the following:

1. If a disposable type has a finalizer and the consuming code _neglects_ to call `Dispose()` on an instance, the instance may survive an additional garbage collection (i.e. work and RAM overhead) and eventually has its unmanaged resources freed. This is better than the alternative without a finalizer, where the unmanaged resources are not freed.

2. If a disposable type has a finalizer and the consuming code _does_ call `Dispose()`, and the type's `Dispose` method [properly calls `GC.SuppressFinalize(this)`](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose#the-dispose-method), then the instance is removed from the finalization queue, thus eliminating any significant overhead, as explained under (1).

In other words, implementing a finalizer for types with unmanaged resources is strictly better: the case where the developer neglects to call `Dispose()` _improves_, and the case where they do call it _does not become worse_.

([This official doc](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose#the-dispose-method) advocates the invocation of `GC.SuppressFinalize(this)` from `Dispose()`. [The `SuppressFinalize`'s doc](https://docs.microsoft.com/en-us/dotnet/api/system.gc.suppressfinalize#remarks) provides further remarks.)

**Comment by normj** [comment]
As a general rule of always adding a `finalizer` whenever we implement `IDisposable` I don't agree with. Particular because makes the experience potentially indeterministic and could create very hard to diagnose issues.
